File: app/src/main/java/com/lol/MainActivity.java
package com.lol;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import com.lol.boring.Event;
import com.lol.boring.Timeline;
import com.lol.boring.UserInfo;

import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import rx.Observable;
import rx.Observer;
import rx.Subscription;
import rx.android.observables.AndroidObservable;
import rx.subscriptions.Subscriptions;
import rx.util.functions.Func1;
import rx.util.functions.Func2;

public class MainActivity extends Activity implements Observer<Timeline> {

    private Adapter adapter;
    private Observable<Timeline> timeline;
    private Subscription subscription;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ListView eventList = (ListView) findViewById(R.id.timeline);
        adapter = new Adapter(this);
        eventList.setAdapter(adapter);

        // create a recurring ticker
        Observable<Long> ticker = Observable.timer(0, 15, TimeUnit.SECONDS);

        // map each of those ticks to an api request
        Observable<List<Event>> events = ticker.flatMap(new Func1<Long, Observable<List<Event>>>() {
            @Override
            public Observable<List<Event>> call(Long requestId) {
                Log.i("meow", "fetching events");
                Api api = new Api();
                return api.events();
            }
        });

        // load the user info from disk
        Observable<UserInfo> userInfo = new Api().userInfo(this);

        // i'm not sure this work 100% like i expect, but when the user info and the network calls complete, combine the two
        Observable<List<Event>> eventsWithReadInfo = events.zip(userInfo, new Func2<List<Event>, UserInfo, List<Event>>() {
            @Override
            public List<Event> call(List<Event> events, UserInfo user) {
                for (Event e : events) {
                    e.read = user.hasRead(e.id);
                }
                return events;
            }
        });

        // when a list of events comes in, partition them into past and future and emit a timeline
        Observable<Timeline> timeline = events.flatMap(new Func1<List<Event>, Observable<Timeline>>() {
            @Override
            public Observable<Timeline> call(final List<Event> events) {
                return Observable.create(new Observable.OnSubscribeFunc<Timeline>() {
                    @Override
                    public Subscription onSubscribe(Observer<? super Timeline> observer) {
                        Log.i("meow", "partitioning events");
                        final List<Event> newEvents = new ArrayList<Event>();
                        final List<Event> pastEvents = new ArrayList<Event>();
                        final DateTime fakeLastSeenTime = DateTime.now();

                        for (Event e : events) {
                            if (e.occursAt.isAfter(fakeLastSeenTime)) {
                                newEvents.add(e);
                            } else {
                                pastEvents.add(e);
                            }
                        }
                        observer.onNext(new Timeline(pastEvents, newEvents));
//                        observer.onCompleted();
                        return Subscriptions.empty();
                    }
                });
            }
        });

        // create an observable timeline that can deal with an activity's lifecycle
        this.timeline = AndroidObservable.fromActivity(this, timeline);
    }

    @Override protected void onResume() {
        super.onResume();
        Log.i("meow", "subscribing to event stream");
        subscription = timeline.subscribe(this);
    }

    @Override protected void onPause() {
        subscription.unsubscribe();
        Log.i("meow", "unsubscribed from event stream");
        super.onPause();
    }

    @Override
    public void onNext(Timeline timeline) {
        Log.i("meow", "updating ui");
        Log.i("meow", String.format("first event: %s", timeline.pastEvents.get(0)));
        adapter.clear();
        adapter.addAll(timeline.pastEvents);
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onCompleted() {
        // won't ever happen.
        Log.i("meow", "done.");
    }

    @Override
    public void onError(Throwable throwable) {
        Log.e("meow", Log.getStackTraceString(throwable));
    }

    private static class Adapter extends ArrayAdapter<Event> {

        public Adapter(Context context) {
            super(context, android.R.layout.simple_list_item_1, android.R.id.text1, new ArrayList<Event>(0));
        }
    }

}

File: app/src/main/kotlin/com/lol/MainActivity.kt
package com.lol

import android.app.Activity
import android.os.Bundle
import com.lol.boring.Event
import rx.Observer
import android.widget.ListView
import android.widget.ArrayAdapter
import java.util.ArrayList
import rx.Observable
import java.util.concurrent.TimeUnit
import java.util.TreeSet
import android.util.Log
import rx.android.observables.AndroidObservable
import rx.Subscription

public open class MainActivity : Activity(), Observer<Set<Event>> {
    private var adapter: ArrayAdapter<Event>? = null
    private var observableTimeline: Observable<TreeSet<Event>>? = null
    private var subscription: Subscription? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super<Activity>.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        adapter = ArrayAdapter<Event>(this, android.R.layout.simple_list_item_1, android.R.id.text1, ArrayList<Event>());
        val listView = findViewById(R.id.timeline) as ListView
        listView.setAdapter(adapter)

        // create an observable set of current events every 15s
        val events = Observable.timer(0, 15, TimeUnit.SECONDS)
        ?.flatMap { Api().events() } // map each tick to an api call.
        ?.filter { it?.occursAt?.isBeforeNow() } // filter out future events
        ?.scan(TreeSet<Event>(), {(es: TreeSet<Event>?, e: Event?) -> if (e != null) es?.add(e as Event); es }) // do a bunch of kotlin bookkeeping and put the event in the set
        ?.debounce(1, TimeUnit.SECONDS) // don't flood the UI with udpates

        // activity lifecycle aware observable, will subscribe on a thread pool and observe on the ui thread
        observableTimeline = AndroidObservable.fromActivity(this, events)
    }

    override fun onResume() {
        super<Activity>.onResume()
        subscription = observableTimeline?.subscribe(this)
    }

    override fun onPause() {
        subscription?.unsubscribe()
        super<Activity>.onPause()
    }

    override fun onNext(events: Set<Event>?) {
        Log.i("meow", "updating ui")
        adapter?.clear()
        adapter?.addAll(events)
        adapter?.notifyDataSetChanged()
    }

    override fun onCompleted() {
        Log.i("meow", "complete.")
    }

    override fun onError(error: Throwable?) {
        Log.e("meow", Log.getStackTraceString(error))
    }

}